README.md
# RequestMaster - Simplified HTTP/HTTPS client
[![npm package](https://nodei.co/npm/request-master.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/request-master/)
[![Build status](https://img.shields.io/travis/request/request/master.svg?style=flat-square)](https://travis-ci.org/request/request)
[![Coverage](https://img.shields.io/codecov/c/github/request/request.svg?style=flat-square)](https://codecov.io/github/request/request?branch=master)
[![Coverage](https://img.shields.io/coveralls/request/request.svg?style=flat-square)](https://coveralls.io/r/request/request)
[![Dependency Status](https://img.shields.io/david/thejavaguy/request-master.svg?style=flat-square)](https://david-dm.org/thejavaguy/request-master)
[![Known Vulnerabilities](https://snyk.io/test/npm/request-master/badge.svg?style=flat-square)](https://snyk.io/test/npm/request-master)

# Under active development!
RequestMaster is a drop-in, API compatible replacement for the original Request! Original Request library [is deprecated](https://github.com/request/request/issues/3142) as of Feb 11th 2020. But don't worry, **RequestMaster is actively developed**! We'll deliver new releases once a quarter, starting from Q4 2020.

## Super simple to use
RequestMaster is designed to be the simplest way possible to make HTTP and HTTPS calls. It follows redirects by default.

```js
const request = require('request-master');
request('https://thejavaguy.org', function onResponse(error, response, body) {
  console.error('error:', error); // Print the error if one occurred
  console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
  console.log('body:', body); // Print the HTML
});
```

## Table of contents
- [Streaming](#streaming)
- [Promises & Async/Await](#promises--asyncawait)
- [Forms](#forms)
- [HTTP Authentication](#http-authentication)
- [Custom HTTP Headers](#custom-http-headers)
- [OAuth Signing](#oauth-signing)
- [Proxies](#proxies)
- [UNIX Domain Sockets](#unix-domain-sockets)
- [TLS/SSL Protocol](#tlsssl-protocol)
- [Support for HAR 1.2](#support-for-har-12)
- [All Available Options](#all-available-options)
- [Convenience methods](#convenience-methods)
- [Debugging](#debugging)
- [Timeouts](#timeouts)
- [Examples](#examples)

---

## Streaming
You can stream any response to a file stream.

```js
request('https://mysite.com/image.png').pipe(fs.createWriteStream('image.png'))
```

You can also stream a file to a PUT or POST request. This method will also check the file extension against a mapping of file extensions to content-types (in this case `application/json`) and use the proper `content-type` in the PUT request (if the headers don’t already provide one).

```js
fs.createReadStream('file.json').pipe(request.put('http://mysite.com/obj.json'))
```

RequestMaster can also `pipe` to itself. When doing so, `content-type` and `content-length` are preserved in the PUT headers.

```js
request.get('https://duckduckgo.com/img.png').pipe(request.put('http://mysite.com/img.png'))
```

RequestMaster emits a "response" event when a response is received. The `response` argument will be an instance of [http.IncomingMessage](https://nodejs.org/api/http.html#http_class_http_incomingmessage).

```js
request
  .get('https://duckduckgo.com/img.png')
  .on('response', function(response) {
    console.log(response.statusCode) // 200
    console.log(response.headers['content-type']) // 'image/png'
  })
  .pipe(request.put('http://mysite.com/img.png'))
```

To easily handle errors when streaming requests, listen to the `error` event before piping:

```js
request
  .get('http://mysite.com/image.png')
  .on('error', function(err) {
    console.error(err)
  })
  .pipe(fs.createWriteStream('image.png'))
```

Now let’s get fancy.

```js
http.createServer(function (req, resp) {
  if (req.url === '/image.png') {
    if (req.method === 'PUT') {
      req.pipe(request.put('http://mysite.com/image.png'))
    } else if (req.method === 'GET' || req.method === 'HEAD') {
      request.get('http://mysite.com/image.png').pipe(resp)
    }
  }
})
```

You can also `pipe()` from `http.ServerRequest` instances, as well as to `http.ServerResponse` instances. The HTTP method, headers, and entity-body data will be sent. Which means that, if you don't really care about security, you can do:

```js
http.createServer(function (req, resp) {
  if (req.url === '/image.png') {
    const x = request('http://mysite.com/image.png')
    req.pipe(x)
    x.pipe(resp)
  }
})
```

And since `pipe()` returns the destination stream in ≥ Node 0.5.x you can do one line proxying.

```js
req.pipe(request('http://mysite.com/image.png')).pipe(resp)
```

Also, none of this new functionality conflicts with requests previous features, it just expands them.

```js
const r = request.defaults({'proxy':'http://localproxy.com'})

http.createServer(function (req, resp) {
  if (req.url === '/image.png') {
    r.get('https://duckduckgo.com/image.png').pipe(resp)
  }
})
```

You can still use intermediate proxies, the requests will still follow HTTP forwards, etc.

[back to top](#table-of-contents)

---

## Promises & Async/Await
RequestMaster supports both streaming and callback interfaces natively. If you'd like RequestMaster to return a Promise instead, you can use an alternative interface wrapper. These wrappers can be useful if you prefer to work with Promises, or if you'd like to use `async`/`await` in ES2017.

Several alternative interfaces are provided by the original request team, including:
- [`request-promise`](https://github.com/request/request-promise) (uses [Bluebird](https://github.com/petkaantonov/bluebird) Promises)
- [`request-promise-native`](https://github.com/request/request-promise-native) (uses native Promises)
- [`request-promise-any`](https://github.com/request/request-promise-any) (uses [any-promise](https://www.npmjs.com/package/any-promise) Promises)

Also, [`util.promisify`](https://nodejs.org/api/util.html#util_util_promisify_original), which is available from Node.js v8.0 can be used to convert a regular function that takes a callback to return a promise instead.

[back to top](#table-of-contents)

---

## Forms
RequestMaster supports `application/x-www-form-urlencoded` and `multipart/form-data` form uploads. For `multipart/related` refer to the `multipart` API.

#### application/x-www-form-urlencoded (URL-Encoded Forms)
URL-encoded forms are simple.

```js
request.post('http://service.com/upload', {form:{key:'value'}})
// or
request.post('http://service.com/upload').form({key:'value'})
// or
request.post({url:'http://service.com/upload', form: {key:'value'}}, function(err,httpResponse,body){ /* ... */ })
```

#### multipart/form-data (Multipart Form Uploads)
For `multipart/form-data` we use the [form-data](https://github.com/form-data/form-data) library by [@felixge](https://github.com/felixge). For the most cases, you can pass your upload form data via the `formData` option.

```js
const formData = {
  // Pass a simple key-value pair
  my_field: 'my_value',
  // Pass data via Buffers
  my_buffer: Buffer.from([1, 2, 3]),
  // Pass data via Streams
  my_file: fs.createReadStream(__dirname + '/unicycle.jpg'),
  // Pass multiple values /w an Array
  attachments: [
    fs.createReadStream(__dirname + '/attachment1.jpg'),
    fs.createReadStream(__dirname + '/attachment2.jpg')
  ],
  // Pass optional meta-data with an 'options' object with style: {value: DATA, options: OPTIONS}
  // Use case: for some types of streams, you'll need to provide "file"-related information manually.
  // See the `form-data` README for more information about options: https://github.com/form-data/form-data
  custom_file: {
    value:  fs.createReadStream('/dev/urandom'),
    options: {
      filename: 'topsecret.jpg',
      contentType: 'image/jpeg'
    }
  }
};
request.post({url:'http://service.com/upload', formData: formData}, function optionalCallback(err, httpResponse, body) {
  if (err) {
    return console.error('upload failed:', err);
  }
  console.log('Upload successful!  Server responded with:', body);
});
```

For advanced cases, you can access the form-data object itself via `r.form()`. This can be modified until the request is fired on the next cycle of the event-loop. (Note that this calling `form()` will clear the currently set form data for that request.)

```js
// NOTE: Advanced use-case, for normal use see 'formData' usage above
const r = request.post('http://service.com/upload', function optionalCallback(err, httpResponse, body) {...})
const form = r.form();
form.append('my_field', 'my_value');
form.append('my_buffer', Buffer.from([1, 2, 3]));
form.append('custom_file', fs.createReadStream(__dirname + '/unicycle.jpg'), {filename: 'unicycle.jpg'});
```
See the [form-data README](https://github.com/form-data/form-data) for more information & examples.

#### multipart/related
Some variations in different HTTP implementations require a newline/CRLF before, after, or both before and after the boundary of a `multipart/related` request (using the multipart option). This has been observed in the .NET WebAPI version 4.0. You can turn on a boundary preambleCRLF or postamble by passing them as `true` to your request options.

```js
request({
  method: 'PUT',
  preambleCRLF: true,
  postambleCRLF: true,
  uri: 'http://service.com/upload',
  multipart: [
    {
      'content-type': 'application/json',
      body: JSON.stringify({foo: 'bar', _attachments: {'message.txt': {follows: true, length: 18, 'content_type': 'text/plain' }}})
    },
    { body: 'I am an attachment' },
    { body: fs.createReadStream('image.png') }
  ],
  // alternatively pass an object containing additional options
  multipart: {
    chunked: false,
    data: [
      {
        'content-type': 'application/json',
        body: JSON.stringify({foo: 'bar', _attachments: {'message.txt': {follows: true, length: 18, 'content_type': 'text/plain' }}})
      },
      { body: 'I am an attachment' }
    ]
  }
},
function (error, response, body) {
  if (error) {
    return console.error('upload failed:', error);
  }
  console.log('Upload successful!  Server responded with:', body);
})
```

[back to top](#table-of-contents)

---

## HTTP Authentication
```js
request.get('http://some.server.com/').auth('username', 'password', false);
// or
request.get('http://some.server.com/', {
  'auth': {
    'user': 'username',
    'pass': 'password',
    'sendImmediately': false
  }
});
// or
request.get('http://some.server.com/').auth(null, null, true, 'bearerToken');
// or
request.get('http://some.server.com/', {
  'auth': {
    'bearer': 'bearerToken'
  }
});
```

If passed as an option, `auth` should be a hash containing values:

- `user` || `username`
- `pass` || `password`
- `sendImmediately` (optional)
- `bearer` (optional)

The method form takes parameters
`auth(username, password, sendImmediately, bearer)`.

`sendImmediately` defaults to `true`, which causes a basic or bearer authentication header to be sent. If `sendImmediately` is `false`, then `request` will retry with a proper authentication header after receiving a `401` response from the server (which must contain a `WWW-Authenticate` header indicating the required authentication method).

Note that you can also specify basic authentication using the URL itself, as detailed in [RFC 1738](http://www.ietf.org/rfc/rfc1738.txt). Simply pass the `user:password` before the host with an `@` sign:

```js
const username = 'username',
    password = 'password',
    url = 'http://' + username + ':' + password + '@some.server.com';

request({url}, function onResponse(error, response, body) {
   // Do more stuff with 'body' here
});
```

Digest authentication is supported, but it only works with `sendImmediately` set to `false`; otherwise `request` will send basic authentication on the initial request, which will probably cause the request to fail.

Bearer authentication is supported, and is activated when the `bearer` value is available. The value may be either a `String` or a `Function` returning a `String`. Using a function to supply the bearer token is particularly useful if used in conjunction with `defaults` to allow a single function to supply the last known token at the time of sending a request, or to compute one on the fly.

[back to top](#table-of-contents)

---

## Custom HTTP Headers
HTTP Headers, such as `User-Agent`, can be set in the `options` object. In the example below, we call the github API to find out the number of stars and forks for the request repository. This requires a custom `User-Agent` header as well as https.

```js
const request = require('request-master');

const options = {
  url: 'https://api.github.com/repos/TheJavaGuy/request-master',
  headers: {
    'User-Agent': 'request-master'
  }
};

function callback(error, response, body) {
  if (!error && response.statusCode == 200) {
    const info = JSON.parse(body);
    console.log(info.stargazers_count + " Stars");
    console.log(info.forks_count + " Forks");
  }
}

request(options, callback);
```

[back to top](#table-of-contents)

---

## OAuth Signing
[OAuth version 1.0](https://tools.ietf.org/html/rfc5849) is supported. The default signing algorithm is [HMAC-SHA1](https://tools.ietf.org/html/rfc5849#section-3.4.2):

```js
// OAuth1.0 - 3-legged server side flow (Twitter example)
// step 1
const qs = require('querystring')
  , oauth =
    { callback: 'http://mysite.com/callback/'
    , consumer_key: CONSUMER_KEY
    , consumer_secret: CONSUMER_SECRET
    }
  , url = 'https://api.twitter.com/oauth/request_token'
  ;
request.post({url:url, oauth:oauth}, function onResponse(e, r, body) {
  // Ideally, you would take the body in the response
  // and construct a URL that a user clicks on (like a sign in button).
  // The verifier is only available in the response after a user has
  // verified with twitter that they are authorizing your app.

  // step 2
  const req_data = qs.parse(body)
  const uri = 'https://api.twitter.com/oauth/authenticate'
    + '?' + qs.stringify({oauth_token: req_data.oauth_token})
  // redirect the user to the authorize uri

  // step 3
  // after the user is redirected back to your server
  const auth_data = qs.parse(body)
    , oauth =
      { consumer_key: CONSUMER_KEY
      , consumer_secret: CONSUMER_SECRET
      , token: auth_data.oauth_token
      , token_secret: req_data.oauth_token_secret
      , verifier: auth_data.oauth_verifier
      }
    , url = 'https://api.twitter.com/oauth/access_token'
    ;
  request.post({url:url, oauth:oauth}, function (e, r, body) {
    // ready to make signed requests on behalf of the user
    const perm_data = qs.parse(body)
      , oauth =
        { consumer_key: CONSUMER_KEY
        , consumer_secret: CONSUMER_SECRET
        , token: perm_data.oauth_token
        , token_secret: perm_data.oauth_token_secret
        }
      , url = 'https://api.twitter.com/1.1/users/show.json'
      , qs =
        { screen_name: perm_data.screen_name
        , user_id: perm_data.user_id
        }
      ;
    request.get({url:url, oauth:oauth, qs:qs, json:true}, function (e, r, user) {
      console.log(user)
    })
  })
})
```

For [RSA-SHA1 signing](https://tools.ietf.org/html/rfc5849#section-3.4.3), make the following changes to the OAuth options object:
* Pass `signature_method : 'RSA-SHA1'`
* Instead of `consumer_secret`, specify a `private_key` string in [PEM format](https://geeklah.com/working_with_pem_files.html)

For [PLAINTEXT signing](https://oauth.net/core/1.0/#anchor22), make the following changes to the OAuth options object:
* Pass `signature_method : 'PLAINTEXT'`

To send OAuth parameters via query params or in a post body as described in The [Consumer Request Parameters](https://oauth.net/core/1.0/#consumer_req_param) section of the oauth1 spec:
* Pass `transport_method : 'query'` or `transport_method : 'body'` in the OAuth options object.
* `transport_method` defaults to `'header'`

To use [Request Body Hash](https://oauth.googlecode.com/svn/spec/ext/body_hash/1.0/oauth-bodyhash.html) you can either
* Manually generate the body hash and pass it as a string `body_hash: '...'`
* Automatically generate the body hash by passing `body_hash: true`

[back to top](#table-of-contents)

---

## Proxies
If you specify a `proxy` option, then the request (and any subsequent redirects) will be sent via a connection to the proxy server.

If your endpoint is an `https` URL, and you are using a proxy, then request will send a `CONNECT` request to the proxy server *first*, and then use the supplied connection to connect to the endpoint.

That is, first it will make a request like:

```
HTTP/1.1 CONNECT endpoint-server.com:80
Host: proxy-server.com
User-Agent: whatever user agent you specify
```

and then the proxy server make a TCP connection to `endpoint-server` on port `80`, and return a response that looks like:

```
HTTP/1.1 200 OK
```

At this point, the connection is left open, and the client is communicating directly with the `endpoint-server.com` machine.

See [the wikipedia page on HTTP Tunneling](https://en.wikipedia.org/wiki/HTTP_tunnel) for more information.

By default, when proxying `http` traffic, request will simply make a standard proxied `http` request. This is done by making the `url` section of the initial line of the request a fully qualified URL to the endpoint.

For example, it will make a single request that looks like:

```
HTTP/1.1 GET http://endpoint-server.com/some-url
Host: proxy-server.com
Other-Headers: all go here

request body or whatever
```

Because a pure "HTTP over HTTP" tunnel offers no additional security or other features, it is generally simpler to go with a straightforward HTTP proxy in this case. However, if you would like to force a tunneling proxy, you may set the `tunnel` option to `true`.

You can also make a standard proxied `http` request by explicitly setting `tunnel : false`, but **note that this will allow the proxy to see the traffic to/from the destination server**.

If you are using a tunneling proxy, you may set the `proxyHeaderWhiteList` to share certain headers with the proxy.

You can also set the `proxyHeaderExclusiveList` to share certain headers only with the proxy and not with destination host.

By default, this set is:

```
accept
accept-charset
accept-encoding
accept-language
accept-ranges
cache-control
content-encoding
content-language
content-length
content-location
content-md5
content-range
content-type
connection
date
expect
max-forwards
pragma
proxy-authorization
referer
te
transfer-encoding
user-agent
via
```

Note that, when using a tunneling proxy, the `proxy-authorization` header and any headers from custom `proxyHeaderExclusiveList` are *never* sent to the endpoint server, but only to the proxy server.

### Controlling proxy behaviour using environment variables
The following environment variables are respected by RequestMaster:

 * `HTTP_PROXY` / `http_proxy`
 * `HTTPS_PROXY` / `https_proxy`
 * `NO_PROXY` / `no_proxy`

When `HTTP_PROXY` / `http_proxy` are set, they will be used to proxy non-SSL requests that do not have an explicit `proxy` configuration option present. Similarly, `HTTPS_PROXY` / `https_proxy` will be respected for SSL requests that do not have an explicit `proxy` configuration option. It is valid to define a proxy in one of the environment variables, but then override it for a specific request, using the `proxy` configuration option. Furthermore, the `proxy` configuration option can be explicitly set to false / null to opt out of proxying altogether for that request.

RequestMaster is also aware of the `NO_PROXY`/`no_proxy` environment variables. These variables provide a granular way to opt out of proxying, on a per-host basis. It should contain a comma separated list of hosts to opt out of proxying. It is also possible to opt of proxying when a particular destination port is used. Finally, the variable may be set to `*` to opt out of the implicit proxy configuration of the other environment variables.

Here are some examples of valid `no_proxy` values:

 * `duckduckgo.com` - don't proxy HTTP/HTTPS requests to DuckDuckGo.
 * `duckduckgo.com:443` - don't proxy HTTPS requests to DuckDuckGo, but *do* proxy HTTP requests to DuckDuckGo.
 * `duckduckgo.com:443, yahoo.com:80` - don't proxy HTTPS requests to DuckDuckGo, and don't proxy HTTP requests to Yahoo!
 * `*` - ignore `https_proxy`/`http_proxy` environment variables altogether.

[back to top](#table-of-contents)

---

## UNIX Domain Sockets
RequestMaster supports making requests to [UNIX Domain Sockets](https://en.wikipedia.org/wiki/Unix_domain_socket). To make one, use the following URL scheme:

```js
/* Pattern */ 'http://unix:SOCKET:PATH'
/* Example */ request.get('http://unix:/absolute/path/to/unix.socket:/request/path')
```

Note: The `SOCKET` path is assumed to be absolute to the root of the host file system.

[back to top](#table-of-contents)

---

## TLS/SSL Protocol
TLS/SSL Protocol options, such as `cert`, `key` and `passphrase`, can be set directly in `options` object, in the `agentOptions` property of the `options` object, or even in `https.globalAgent.options`. Keep in mind that, although `agentOptions` allows for a slightly wider range of configurations, the recommended way is via `options` object directly, as using `agentOptions` or `https.globalAgent.options` would not be applied in the same way in proxied environments (as data travels through a TLS connection instead of an http/https agent).

```js
const fs = require('fs')
    , path = require('path')
    , certFile = path.resolve(__dirname, 'ssl/client.crt')
    , keyFile = path.resolve(__dirname, 'ssl/client.key')
    , caFile = path.resolve(__dirname, 'ssl/ca.cert.pem')
    , request = require('request-master');

const options = {
  url: 'https://api.some-server.com/',
  cert: fs.readFileSync(certFile),
  key: fs.readFileSync(keyFile),
  passphrase: 'password',
  ca: fs.readFileSync(caFile)
};

request.get(options);
```

### Using `options.agentOptions`
In the example below, we call an API that requires client side SSL certificate
(in PEM format) with passphrase protected private key (in PEM format) and disable the SSLv3 protocol:

```js
const fs = require('fs')
    , path = require('path')
    , certFile = path.resolve(__dirname, 'ssl/client.crt')
    , keyFile = path.resolve(__dirname, 'ssl/client.key')
    , request = require('request-master');

const options = {
  url: 'https://api.some-server.com/',
  agentOptions: {
    cert: fs.readFileSync(certFile),
    key: fs.readFileSync(keyFile),
    // Or use `pfx` property replacing `cert` and `key` when using private key, certificate and CA certs in PFX or PKCS12 format:
    // pfx: fs.readFileSync(pfxFilePath),
    passphrase: 'password',
    securityOptions: 'SSL_OP_NO_SSLv3'
  }
};

request.get(options);
```

It is able to force using SSLv3 only by specifying `secureProtocol`:

```js
request.get({
  url: 'https://api.some-server.com/',
  agentOptions: {
    secureProtocol: 'SSLv3_method'
  }
});
```

It is possible to accept other certificates than those signed by generally allowed Certificate Authorities (CAs). This can be useful, for example,  when using self-signed certificates. To require a different root certificate, you can specify the signing CA by adding the contents of the CA's certificate file to the `agentOptions`. The certificate the domain presents must be signed by the root certificate specified:

```js
request.get({
  url: 'https://api.some-server.com/',
  agentOptions: {
    ca: fs.readFileSync('ca.cert.pem')
  }
});
```

The `ca` value can be an array of certificates, in the event you have a private or internal corporate public-key infrastructure hierarchy. For example, if you want to connect to https://api.some-server.com which presents a key chain consisting of:

1. its own public key, which is signed by:
2. an intermediate "Corp Issuing Server", that is in turn signed by:
3. a root CA "Corp Root CA";

you can configure your request as follows:

```js
request.get({
  url: 'https://api.some-server.com/',
  agentOptions: {
    ca: [
      fs.readFileSync('Corp Issuing Server.pem'),
      fs.readFileSync('Corp Root CA.pem')
    ]
  }
});
```

[back to top](#table-of-contents)

---

## Support for HAR 1.2
The `options.har` property will override the values: `url`, `method`, `qs`, `headers`, `form`, `formData`, `body`, `json`, as well as construct multipart data and read files from disk when `request.postData.params[].fileName` is present without a matching `value`.

A validation step will check if the HAR Request format matches the latest spec (v1.2) and will skip parsing if not matching.

```js
const request = require('request-master')
request({
  // will be ignored
  method: 'GET',
  uri: 'https://duckduckgo.com',

  // HTTP Archive Request Object
  har: {
    url: 'http://www.mockbin.com/har',
    method: 'POST',
    headers: [
      {
        name: 'content-type',
        value: 'application/x-www-form-urlencoded'
      }
    ],
    postData: {
      mimeType: 'application/x-www-form-urlencoded',
      params: [
        {
          name: 'foo',
          value: 'bar'
        },
        {
          name: 'hello',
          value: 'world'
        }
      ]
    }
  }
})

// a POST request will be sent to http://www.mockbin.com
// with body an application/x-www-form-urlencoded body:
// foo=bar&hello=world
```

[back to top](#table-of-contents)

---

## All available options
`request(options, callback)`


CHANGELOG.md
## Change Log

### v3.0.0 '(2021/03/11)'
- [#21](https://github.com/TheJavaGuy/request-master/pull/21) replace phantomjs with headless browser (@TheJavaGuy)
- [#20](https://github.com/TheJavaGuy/request-master/pull/20) switch to CircleCI (@TheJavaGuy)
- [#19](https://github.com/TheJavaGuy/request-master/pull/19) add docker container for tests (@TheJavaGuy)

### v2.89.0 '(2020/12/15)'
- [#16](https://github.com/TheJavaGuy/request-master/pull/16) update karma (@TheJavaGuy)
- [#15](https://github.com/TheJavaGuy/request-master/pull/15) add PR template (@TheJavaGuy)
- [#12](https://github.com/TheJavaGuy/request-master/pull/12) update http-signature to 1.3.5 (@TheJavaGuy)
- [#10](https://github.com/TheJavaGuy/request-master/pull/10) add node.js v12 to TravisCI (@TheJavaGuy)
- [#7](https://github.com/TheJavaGuy/request-master/pull/7) switch from standard to semistandard (@TheJavaGuy)
- [#4](https://github.com/TheJavaGuy/request-master/pull/4) update har-validator to 5.1.5 (@TheJavaGuy)
- [#3268](https://github.com/request/request/pull/3268) README: typo in deprecation message (#3268) (@gr2m)
- [#3267](https://github.com/request/request/pull/3267) doc: note full deprecation (#3267) (@mikeal)
- [#3169](https://github.com/request/request/pull/3169) Use nyc instead of istanbul (#3169) (@andresroberto)
- [#3023](https://github.com/request/request/pull/3023) Update README.md (#3023) (@odykyi)
- [#3008](https://github.com/request/request/pull/3008) docs: :memo: Updated the Change Log for v2.88.0 (2018/08/10) (#3008) (@shige)
- [#3115](https://github.com/request/request/pull/3115) docs: adjust readme timeout argument description to be clearer (#3115) (@jaredsuttles)
- [#3109](https://github.com/request/request/pull/3109) Update outdated link in comment (#3109) (@flotwig)
- [#3090](https://github.com/request/request/pull/3090) fix: Clear timeout on `#abort()` (#3090) (@reconbot)
- [#3089](https://github.com/request/request/pull/3089) chore: Update to tough-cookie 2.5.0 (#3089) (@grabus)
- [#2972](https://github.com/request/request/pull/2972) optimize:do unnecessary processing when changing options to null (#2972) (@rickyes)
- [#3074](https://github.com/request/request/pull/3074) fix: expression minor bugs (#3074) (@A1vinSmith)
- [#2808](https://github.com/request/request/pull/2808) Revert "Workaround for GitHub issue #2807 (#2808)" (#3075) (@reconbot)
- [#2808](https://github.com/request/request/pull/2808) Workaround for GitHub issue #2807 (#2808) (@pgonzal)
- [#3064](https://github.com/request/request/pull/3064) chore: ensure piped body is exactly the same (#3064) (@reconbot)
- [#3046](https://github.com/request/request/pull/3046) fix: use const instead of var in the README (#3046) (@EmmanuelDemey)
- [#2995](https://github.com/request/request/pull/2995) Documented response.caseless.get() (#2995) (@lifeisfoo)
- [#2701](https://github.com/request/request/pull/2701) Verify that body is not empty (#2701) (@RolfKoenders)
- [#2728](https://github.com/request/request/pull/2728) Update the documentation of timeout to reflect it controls timeout of socket inactivity (#2728) (@odedpeer)
- [#2750](https://github.com/request/request/pull/2750) Update cookie store API link (#2750) (@Huachao)
- [#2824](https://github.com/request/request/pull/2824) Update README to mention util.promisify() (#2824) (@riverspirit)
- [#2674](https://github.com/request/request/pull/2674) Fix link to HAR 1.2 Section Anchor (#2674) (@Maciek416)
- [#2644](https://github.com/request/request/pull/2644) Update README.md (#2644) (@deed02392)
- [#3062](https://github.com/request/request/pull/3062) fix: SSL Certs (#3062) (@reconbot)
- [#3061](https://github.com/request/request/pull/3061) Add neverstale label to stalebot config (@reconbot)
- [#3060](https://github.com/request/request/pull/3060) chore: Add probot-stale (@reconbot)
- [#3055](https://github.com/request/request/pull/3055) fix #3054, Bump har-validator version to 5.1.3 (#3055) (@JChanceHud)

### v2.88.0 (2018/08/10)
- [#2996](https://github.com/request/request/pull/2996) fix(uuid): import versioned uuid (@kwonoj)
- [#2994](https://github.com/request/request/pull/2994) Update to oauth-sign 0.9.0 (@dlecocq)
- [#2993](https://github.com/request/request/pull/2993) Fix header tests (@simov)
- [#2904](https://github.com/request/request/pull/2904) #515, #2894 Strip port suffix from Host header if the protocol is known. (#2904) (@paambaati)
- [#2791](https://github.com/request/request/pull/2791) Improve AWS SigV4 support. (#2791) (@vikhyat)
- [#2977](https://github.com/request/request/pull/2977) Update test certificates (@simov)

### v2.87.0 (2018/05/21)
- [#2943](https://github.com/request/request/pull/2943) Replace hawk dependency with a local implemenation (#2943) (@hueniverse)

### v2.86.0 (2018/05/15)
- [#2885](https://github.com/request/request/pull/2885) Remove redundant code (for Node.js 0.9.4 and below) and dependency (@ChALkeR)
- [#2942](https://github.com/request/request/pull/2942) Make Test GREEN Again! (@simov)
- [#2923](https://github.com/request/request/pull/2923) Alterations for failing CI tests (@gareth-robinson)

### v2.85.0 (2018/03/12)
- [#2880](https://github.com/request/request/pull/2880) Revert "Update hawk to 7.0.7 (#2880)" (@simov)

### v2.84.0 (2018/03/12)
- [#2793](https://github.com/request/request/pull/2793) Fixed calculation of oauth_body_hash, issue #2792 (@dvishniakov)
- [#2880](https://github.com/request/request/pull/2880) Update hawk to 7.0.7 (#2880) (@kornel-kedzierski)

### v2.83.0 (2017/09/27)
- [#2776](https://github.com/request/request/pull/2776) Updating tough-cookie due to security fix. (#2776) (@karlnorling)

### v2.82.0 (2017/09/19)
- [#2703](https://github.com/request/request/pull/2703) Add Node.js v8 to Travis CI (@ryysud)
- [#2751](https://github.com/request/request/pull/2751) Update of hawk and qs to latest version (#2751) (@Olivier-Moreau)

lib/tunnel.js
'use strict';

var url = require('url');
var tunnel = require('tunnel-agent');

var defaultProxyHeaderWhiteList = [
  'accept',
  'accept-charset',
  'accept-encoding',
  'accept-language',
  'accept-ranges',
  'cache-control',
  'content-encoding',
  'content-language',
  'content-location',
  'content-md5',
  'content-range',
  'content-type',
  'connection',
  'date',
  'expect',
  'max-forwards',
  'pragma',
  'referer',
  'te',
  'user-agent',
  'via'
];

var defaultProxyHeaderExclusiveList = [
  'proxy-authorization'
];

function constructProxyHost (uriObject) {
  var port = uriObject.port;
  var protocol = uriObject.protocol;
  var proxyHost = uriObject.hostname + ':';

  if (port) {
    proxyHost += port;
  } else if (protocol === 'https:') {
    proxyHost += '443';
  } else {
    proxyHost += '80';
  }

  return proxyHost;
}

function constructProxyHeaderWhiteList (headers, proxyHeaderWhiteList) {
  var whiteList = proxyHeaderWhiteList
    .reduce(function (set, header) {
      set[header.toLowerCase()] = true;
      return set;
    }, {});

  return Object.keys(headers)
    .filter(function (header) {
      return whiteList[header.toLowerCase()];
    })
    .reduce(function (set, header) {
      set[header] = headers[header];
      return set;
    }, {});
}

function constructTunnelOptions (request, proxyHeaders) {
  var proxy = request.proxy;

  var tunnelOptions = {
    proxy: {
      host: proxy.hostname,
      port: +proxy.port,
      proxyAuth: proxy.auth,
      headers: proxyHeaders
    },
    headers: request.headers,
    ca: request.ca,
    cert: request.cert,
    key: request.key,
    passphrase: request.passphrase,
    pfx: request.pfx,
    ciphers: request.ciphers,
    rejectUnauthorized: request.rejectUnauthorized,
    secureOptions: request.secureOptions,
    secureProtocol: request.secureProtocol
  };

  return tunnelOptions;
}

function constructTunnelFnName (uri, proxy) {
  var uriProtocol = (uri.protocol === 'https:' ? 'https' : 'http');
  var proxyProtocol = (proxy.protocol === 'https:' ? 'Https' : 'Http');
  return [uriProtocol, proxyProtocol].join('Over');
}

function getTunnelFn (request) {
  var uri = request.uri;
  var proxy = request.proxy;
  var tunnelFnName = constructTunnelFnName(uri, proxy);
  return tunnel[tunnelFnName];
}

function Tunnel (request) {
  this.request = request;
  this.proxyHeaderWhiteList = defaultProxyHeaderWhiteList;
  this.proxyHeaderExclusiveList = [];
  if (typeof request.tunnel !== 'undefined') {
    this.tunnelOverride = request.tunnel;
  }
}

Tunnel.prototype.isEnabled = function () {
  var self = this;
  var request = self.request;
  // Tunnel HTTPS by default. Allow the user to override this setting.

  // If self.tunnelOverride is set (the user specified a value), use it.
  if (typeof self.tunnelOverride !== 'undefined') {
    return self.tunnelOverride;
  }

  // If the destination is HTTPS, tunnel.
  if (request.uri.protocol === 'https:') {
    return true;
  }

  // Otherwise, do not use tunnel.
  return false;
};

Tunnel.prototype.setup = function (options) {
  var self = this;
  var request = self.request;

  options = options || {};

  if (typeof request.proxy === 'string') {
    request.proxy = url.parse(request.proxy);
  }

  if (!request.proxy || !request.tunnel) {
    return false;
  }

  // Setup Proxy Header Exclusive List and White List
  if (options.proxyHeaderWhiteList) {
    self.proxyHeaderWhiteList = options.proxyHeaderWhiteList;
  }
  if (options.proxyHeaderExclusiveList) {
    self.proxyHeaderExclusiveList = options.proxyHeaderExclusiveList;
  }

  var proxyHeaderExclusiveList = self.proxyHeaderExclusiveList.concat(defaultProxyHeaderExclusiveList);
  var proxyHeaderWhiteList = self.proxyHeaderWhiteList.concat(proxyHeaderExclusiveList);

  // Setup Proxy Headers and Proxy Headers Host
  // Only send the Proxy White Listed Header names
  var proxyHeaders = constructProxyHeaderWhiteList(request.headers, proxyHeaderWhiteList);
  proxyHeaders.host = constructProxyHost(request.uri);

  proxyHeaderExclusiveList.forEach(request.removeHeader, request);

  // Set Agent from Tunnel Data
  var tunnelFn = getTunnelFn(request);
  var tunnelOptions = constructTunnelOptions(request, proxyHeaders);
  request.agent = tunnelFn(tunnelOptions);

  return true;
};

Tunnel.defaultProxyHeaderWhiteList = defaultProxyHeaderWhiteList;
Tunnel.defaultProxyHeaderExclusiveList = defaultProxyHeaderExclusiveList;
exports.Tunnel = Tunnel;
